### ISSUE TITLE
IncrementalSearchCV runs an adaptive algorithm by default

### ISSUE THREAD

**Issue by stsievert** [dask/dask-ml#388]
There are two dimensions in which model selection searches can vary:

> Does the search...
>
> 1. use `partial_fit` or `fit`?
> 2. use previous evaluations to choose which models to evaluate?
>
> (adapted from https://github.com/dask/dask-ml/pull/370#discussion_r221819839, which is a response to https://github.com/dask/dask-ml/pull/370#discussion_r221573023)

I think (1) should be named "incremental" and (2) should be named "adaptive".

Currently, `IncrementalSearchCV` implements

* a complex adaptive algorithm
* stopping on plateau
* an incremental search that's used with `_incremental.fit`.

I'd like to clean the naming of `IncrementalSearchCV`.

**Comment by mrocklin** [comment]
I think that fundamentally there is a tension here between being precise and being concise.  A name that fully encompassed all of the characteristics of the current implementation probably isn't feasible in less than twenty characters.  It's not clear to me how to make this compromise.  Do you have any thoughts on the right way forward?

**Comment by stsievert** [comment]
> Do you have any thoughts on the right way forward?

I think my biggest complaint is the mixing of the incremental model selection search and the complex adaptive algorithm. Maybe something like

``` python
class IncrementalSearchCV(...):
    def _additional_calls(self, info):
        ...

    def _adapt(self, info):
        # default: don't adapt (allows mirroring RandomizedSearchCV
        # and GridSearchCV, but stopping on plateau)
        return {k: 1 for k in info}

    def _stop_on_plateau(self, info):
        return ...

class TimeDecaySearchCV(IncrementalSearchCV):
    def _adapt(self, info):  # as in #370
        return ...
```

**Comment by mrocklin** [comment]
Is your only concern about the current design that it doesn't allow for a hierarchy of names?  

Originally the plan was to have a super-class that people would inherit from and override `_additional_calls` in order to implement their own solution.  We did this and implemented a couple solutions, but found that it was pretty easy to combine all of the traits that we wanted into a single implementation.  The ability to subclass remains, but the default solution is just decently capable. 

I think I see reducing the pool of models over time to be similar to early stopping or how we select parameters to start with.  They're all features of the same general process.  The extent to which we can make them orthogonal so that they can be changed individually by overloading some method seems nice.  I don't think that any of them need to be pulled out to form a separate class hierarchy to achieve this though.

**Comment by stsievert** [comment]
> Is your only concern about the current design that it doesn't allow for a hierarchy of names?

My concern is that `IncrementalSearchCV` runs a complicated adaptive algorithm by default.

I feel `IncrementalSearchCV` can have the features it provides be better tailored for it's name. `IncrementalSearchCV` is currently providing

1. incremental model selection (repeated calls to `partial_fit`)
2. a basic adaptive algorithm (stop on plateau)
1. a complicated adaptive algorithm (inverse decay)

The adaptive algorithm requires tuning to choose the right level of adaptiveness (by choosing the right values of `n_inital_parameters` and `decay_rate`). The level of adaptiveness that is optimal depends on the problem, model and dataset. This has propagated to the tests: see the conversation at https://github.com/dask/dask-ml/pull/404#discussion_r225218107 and the diff of #406 at [stsievert#2:L237].

I'd rather see `IncrementalSearchCV` provide

1. incremental model selection (repeated calls to `partial_fit`)
2. a basic adaptive algorithm (stop on plateau)
1. ~~a complicated adaptive algorithm (inverse decay)~~ an option to provide an adaptive algorithm (with a passive default).

With this, `IncrementalSearchCV` would mirror `RandomizedSearchCV` or `GridSearchCV` with the default parameters but call `partial_fit` instead of `fit`. It would always return the correct result, even if it takes a long time to compute (but I think we should strongly recommend particular adaptive algorithms in the docs).

[stsievert#2:L237]:https://github.com/stsievert/dask-ml/pull/2/files#diff-7d8385ab7b08e10fef65257ced048e3cR237

**Comment by TomAugspurger** [comment]
I don't see how this discussion on tests in https://github.com/dask/dask-ml/pull/404#discussion_r225218107 is relevant to the default algorithm choice. Those asserts would have been incorrect even if they were manually specified (and we'll want some sanity checks for this, regardless of whether they're the default or not).

> It would always return the correct result, even if it takes a long time to compute (but I think we should strongly recommend particular adaptive algorithms in the docs).

I can see both sides of this. My main concern is in explaining IncrementalSearchCV to people. I worry that if the default mirror RandomaizedSearchCV or GridSearchCV, I'll have difficulty recommending when people should use it vs. `RandomizedSearchCV(Incremental(Estimator), params)`.

**Comment by stsievert** [comment]
> I don't see how this discussion on tests in #404 (comment) is relevant to the default algorithm choice

Mostly, I was pointing to the graph in that comment.

> I can see both sides of this. ... `RandomizedSearchCV(Incremental(Estimator), params)`

Ooo, that's a good argument. I can see both sides now too. Let me sit on this for a bit.

**Comment by stsievert** [comment]
> I'll have difficulty recommending when people should use it vs. `RandomizedSearchCV(Incremental(Est), params)`.

With this, I think my concern is a naming issue.

I see two competing pulls for naming:

1. the requirement: the model must implement `partial_fit`
2. the feature: the search can be adaptive and by default is (even "stop on plateau" is adaptive).

I'm tempted to rename to `IncrementalSearchCV` to `AdaptiveSearchCV`, which fits the design goal: to use previous calls to `partial_fit` to decide whether to continue training. Yes, the class can be configured to be passive, but that's not the focus (and why not use the `Incremental`+Scikit-learn solution above?).

Thoughts?

**Comment by TomAugspurger** [comment]
I think at this point, I'm pretty OK with the name. It's probably just
inertia, but I'm happy to leave it as is.

**Comment by stsievert** [comment]
Renaming to `AdaptiveSearchCV` would close this issue (how can `AdaptiveSearchCV` not be adaptive by default?). I don't think this issue should be closed if `IncrementalSearchCV` persists.

I've opened #412 for this.

**Comment by mrocklin** [comment]
I'm also happy to leave it as is.  Just as incremental search leaves out the adaptive element of the algorithm, adaptive leaves out the incremental element of the name.

**Comment by TomAugspurger** [comment]
I don't think this should block the release (which I'm going to do today). We can easily change the name (and provide a shim for backwards compat) if we want to in the future.

**Comment by stsievert** [comment]
> incremental search leaves out the adaptive element of the algorithm, adaptive leaves out the incremental element of the name.

Why not a name for both? i.e., something like

``` python
class IncrementalSearchCV:
    def _additional_calls(self, info):
        return self._stop_on_plateau(info)

class AdaptiveSearchCV(IncrementalSearchCV):
    def _additional_calls(self, info):
        instructions = self._adapt(info)
        return self._stop_on_plateau(instructions)
```

This cleanly separates the "incremental" and "adaptive" elements.

This is the same implementation as in https://github.com/dask/dask-ml/pull/370#discussion_r224967594, and as discussed there I can make this change in #221. That will provide better motivation and a clear use case.

**Comment by stsievert** [comment]
> I'll have difficulty recommending when people should use it vs. `RandomizedSearchCV(Incremental(Est), params)`.

I think the hyper-parameter search documentation can re-structured to make it clear what Dask-ML offers for hyper-parameter optimization. I've implemented this restructure in #221.

I've changed the default `decay_rate` to 0 in d13384fc63d894a2ba080c291e69d3c6a740e955 as part of #221. This will help `IncrementalSearchCV` serve hyper-parameter searches that memory constrained but not compute constrained.

Functionally, this is the same as `GridSearchCV(Incremental(model), params)` but there are apparently some issues with that: https://github.com/dask/dask-ml/issues/421#issuecomment-435212089.
